### PR TITLE
refactor(scrollable-video): enhance user experience and follow autoplay policy

### PR DESCRIPTION
**Pull request by nickhsine**
# User Experience Goal
Render scrollable video as soon as possible.
The old way is to render scrollable video after video fully downloaded.
In this PR, we change to render scrollable video when first frame of video is ready.

# Additional Changes
1. upload demo video files rather than use twreporter video urls
2. stop `fetch`ing video sources by ourselves. The `fetch` promise will call `then` only if the full data of video is `fetch`ed. But we don't want to wait for downloading all the video data.
3. use `<video>` `src` attribute to load the video source instead.
4. follow autoplay policy and refactor the codes.

NOTE: 
For more Implementation details, please see the commit messages.

### CODE REVIEW

**Comment by nickhsine** [comment]
In case that the playback has stopped because user has scrolled too far, and video lacks of data. I am thinking whether to add `waiting` event listener to show `影片下載中` message or not.
How do you think?

**Comment by taylrj** [comment]
If our goal is to render scrollable video as soon as possible, yes I think so.

Also, I think here is a trade-off between how soon we'd like to let users see the first render of scrollable video and how smooth (I mean, not to be interrupted by any message) we'd like users to experience when the scrollable video plays on scroll and the whole data is not fetched yet.

**Comment by taylrj** [comment]
From where I stand, I'd agree with this change (let users see the first render of scrollable video ASAP) as the use cases could be extended to videos in large file size (e.g. long videos).

I just want to mention that we should know here is a trade-off and we are making the decision in this change.

File: packages/scrollable-video/src/components/video.js [hunk 1]
```diff
@@ -208,8 +208,6 @@ const ForwardRefVideo = React.forwardRef(
     }, [pickedSourceSrc, setVideoError, setVideoLoading, preloadCacheType, ref])
     return (
       <Video
-        playsInline
```
Comment (taylrj) [inline packages/scrollable-video/src/components/video.js:212]: It seems to me that `playsInline` attribute is not for autoplay but [for iOS not to enter fullscreen mode and continue to play inline](https://developer.apple.com/documentation/webkit/delivering_video_content_for_safari#3030250).

From this perspective, I suppose we still need `playsInline` as an attribute for video ?

Comment (taylrj) [reply packages/scrollable-video/src/components/video.js:212]: Oops, I will resolve this one since `playsInline` is added back in next patch.

**Comment by taylrj** [comment]
To mitigate the buffering issue, would we consider to utilize `canplay` instead of `loadeddata` event because

The `loadeddata` event occurs when data for the current frame is loaded, but not enough data to play next frame of the specified audio/video.

But the `canplay` event occurs when the browser can start playing the specified audio/video (when it has buffered enough to begin).

And it is not even close to `canplaythrough` event which occurs when the browser estimates it can play through the specified audio/video without having to stop for buffering.

How do you think ?

**Comment by nickhsine** [comment]
The reason why I use `onloadeddata` rather than `canplay` is because iOS has trouble to fire `canplay` event.
See https://stackoverflow.com/questions/50051639/javascript-html5-video-event-canplay-not-firing-on-safari.
how about we try `onloadeddata` and `waiting` to see what will happen?

**Comment by nickhsine** [comment]
> The reason why I use `onloadeddata` rather than `canplay` is because iOS has trouble to fire `canplay` event. See https://stackoverflow.com/questions/50051639/javascript-html5-video-event-canplay-not-firing-on-safari. how about we try `onloadeddata` and `waiting` to see what will happen?

But I just tried `onCanPlay` event on iOS Safari, it works this time.
I don't know why last time I encountered the problem that `canplay` event was not fired.
Anyway, I will change `onloadeddata` to `canplay` event.
Maybe we can have more tests for `canplay` event.

**Comment by nickhsine** [comment]
[ed0e83e](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/ed0e83e901e272a6ac5998aa14f172bfa0b3d0a2): change `onLoadedData` to `onCanPlay`

**Comment by nickhsine** [comment]
[4d4252d](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/4d4252dd32eeb766c82aabddc7f822030577a02f): add spinner and `onWaiting` event listener

File: packages/scrollable-video/src/components/video.js [hunk 2]
```diff
@@ -185,7 +185,7 @@ const ForwardRefVideo = React.forwardRef(
           true /* `autoPlay` is required by iOS. If `autoPlay` is not true, and then `loadedData` event won't be triggered */
```
Comment (taylrj) [inline packages/scrollable-video/src/components/video.js:185]: If we listen to canplay event, would autoPlay also be required by iOS ?

If so, could we update the comment `loadedData` to `canplay` ?

Comment (nickhsine) [reply packages/scrollable-video/src/components/video.js]: [df4512a](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/df4512af22d8f2d178cce90bca1d5ed5aa26d452) removes `autoPlay` attribute.

Comment (nickhsine) [reply packages/scrollable-video/src/components/video.js]: [df4512a](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/df4512af22d8f2d178cce90bca1d5ed5aa26d452) fixes it.

Comment (taylrj) [reply packages/scrollable-video/src/components/video.js]: Yes it surely does 👍

My last question: 
The fix adds `key` prop to recreate `videoRef` when video source changes.
But I am still curious about why this corner case occurs in this change ?

Before [391928d](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/391928dcc0c6ee2967867a865d852c3ef8cdbb72) and this change, the <Video> did not have `key` prop hence the `videoRef` didn't be recreated and no new ScrollTrigger instances were created by `useScrollTrigger` except for the first one. But why the old ways didn't suffer from this issue ? I still don't get it...

**Comment by taylrj** [comment]
The [iOS corner case fix](https://github.com/twreporter/orangutan-monorepo/pull/130/commits/d2f330eee30fad29f1d1a674e1516b0c379511e2) does work !! 👍

My last question: 

The fix adds `key` prop to recreate `videoRef` when video source changes.
But I am still curious about why this corner case occurs in this change ?
Before this change, the `<Video>` did not have `key` prop hence the `videoRef` wasn't be recreated and no new ScrollTrigger instances were created by `useScrollTrigger` except for the first one. But why the old way didn't suffer from this issue ? I still don't get it...

**Comment by nickhsine** [comment]
> My last question:
> 
> The fix adds `key` prop to recreate `videoRef` when video source changes. But I am still curious about why this corner case occurs in this change ? Before this change, the `<Video>` did not have `key` prop hence the `videoRef` wasn't be recreated and no new ScrollTrigger instances were created by `useScrollTrigger` except for the first one. But why the old way didn't suffer from this issue ? I still don't get it...

If  we change `<Video>` from 
```
<Video
  src={pickedSourceSrc}
>
```
to 
```
<Video
{...(pickedSourceSrc && {src: pickedSourceSrc})}
>
```
, and then we also could fix this iOS bug.

The old program works well is because it pass `src` props to `<Video>` only if `src` is not `undefined`.
It seems that iOS Safari treat `src=""` (empty string) and `src={undefined}` differently.
The latter will make `ref` recreated. 
However,  in previous commits of the PR, we use the former way. Safari does not recreate `ref` even we pass `pickedSourceSrc` after mounted.

**Comment by nickhsine** [comment]
I am not sure why ScrollTrigger can't create right trigger point on iOS Safari. 
It may need more digging.

If the PR is ok, I prefer to merging the PR first, and leave an issue for the above question.
How do you think?

**Review by taylrj — APPROVED** [review]
Ok, maybe we can merge and publish the rc version first.

The cloud-function will be updated only when the stable version of this change is published.
After the test is well performed, we could release the change to stable version.

Thanks, @nickhsine.

**Pull request merged**